src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,224 @@
+ */
+
+#include "mlio/image_reader.h"
+
+#include <fmt/format.h>
+#include <opencv4/opencv2/imgcodecs.hpp>
+#include <opencv4/opencv2/imgproc/imgproc.hpp>
+
+#include "mlio/cpu_array.h"
+#include "mlio/data_type.h"
+#include "mlio/instance.h"
+#include "mlio/instance_batch.h"
+#include "mlio/intrusive_ptr.h"
+#include "mlio/parallel_data_reader.h"
+#include "mlio/record_readers/blob_record_reader.h"
+#include "mlio/record_readers/recordio_record_reader.h"
+#include "mlio/schema.h"
+#include "mlio/streams/utf8_input_stream.h"
+
+namespace mlio {
+inline namespace v1 {
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    if (params_.data_shape.size() != data_shape_size_) {
+        throw std::invalid_argument{
+            "data_shape is a required argument. "
+            "Please enter shape of the output image in "
+            "(channels, height, width) format"};
+    }
+}
+
+image_reader::~image_reader()
+{
+    stop();
+}
+
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &ds)
+{
+    if (params_.img_frame != image_frame::file &&
+        params_.img_frame != image_frame::pipe) {
+        throw std::invalid_argument{"Please enter a valid value for "
+                                    "image_frame parameter."};
+    }
+    switch (params_.img_frame) {
+    case image_frame::file:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::pipe:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+}
+
+void
+image_reader::infer_schema(instance const &)
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(
+        feature_desc_builder{"value",
+                             data_type::uint8,
+                             {params().batch_size,
+                              static_cast<std::size_t>(params_.data_shape[1]),
+                              static_cast<std::size_t>(params_.data_shape[2]),
+                              static_cast<std::size_t>(params_.data_shape[0])}}
+            .build());
+    schema_ = make_intrusive<schema>(std::move(descs));
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &batch) const
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(tsr);
+    return make_intrusive<example>(schema_, std::move(tensors));
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    auto instance_offset =
+        params_.data_shape[0] * params_.data_shape[1] * params_.data_shape[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
```

Comment:
awaiting your comment on my response to your previous comment.

Reply:
"So what I noticed was that making data_shape of type std::size_t lead to more typecasting to int at various places since opencv Mat used int for matrix rows/cols. This way, I was able to achieve cleaner code as number of typecasting decreased?

thoughts?"

Reply:
sounds good. done.